server/src/main/java/io/druid/server/lookup/cache/LookupCoordinatorManager.java

```diff
@@ -506,137 +257,398 @@ public boolean deleteLookup(final String tier, final String lookup, AuditInfo au
+      for (Map.Entry<String, Map<String, LookupExtractorFactoryMapContainer>> tierEntry : allLookupTiers.entrySet()) {
+
+        LOG.debug("Starting lookup mgmt for tier [%s].", tierEntry.getKey());
+
+        final Map<String, LookupExtractorFactoryMapContainer> tierLookups = tierEntry.getValue();
+        for (final HostAndPort node : listenerDiscoverer.getNodes(LookupModule.getTierListenerPath(tierEntry.getKey()))) {
+
+          LOG.debug(
+              "Starting lookup mgmt for tier [%s] and host [%s:%s].",
+              tierEntry.getKey(),
+              node.getHostText(),
+              node.getPort()
+          );
+
+          futures.add(
+              executorService.submit(
+                  () -> this.doLookupManagementOnNode(node, tierLookups, currState)
+              )
+          );
+        }
+      }
+
+      final ListenableFuture allFuture = Futures.allAsList(futures);
+      try {
+        allFuture.get(lookupCoordinatorManagerConfig.getAllHostTimeout().getMillis(), TimeUnit.MILLISECONDS);
+        knownOldState.set(currState);
+      }
+      catch (InterruptedException ex) {
+        allFuture.cancel(true);
+        Thread.currentThread().interrupt();
+        throw ex;
+      } catch (Exception ex) {
+        allFuture.cancel(true);
+        throw ex;
+      }
+
+    } catch (Exception ex) {
+      LOG.makeAlert(ex, "Failed to finish lookup management loop.").emit();
+    }
+
+    LOG.debug("Finished lookup sync for on all nodes.");
+  }
+
+  private void doLookupManagementOnNode(
+      HostAndPort node,
+      Map<String, LookupExtractorFactoryMapContainer> nodeTierLookups,
+      Map<HostAndPort, LookupsStateWithMap> stateToUpdate
+  )
+  {
+    try {
+
+      LOG.debug("Starting lookup sync for node [%s].", node);
+
+      LookupsStateWithMap lookupsState = getLookupStateForNode(node);
+      LOG.debug("Received lookups state from node [%s].", node);
+
+      stateToUpdate.put(node, lookupsState);
+
+      Map<String, LookupExtractorFactoryMapContainer> toLoad = new HashMap<>();
+      for (Map.Entry<String, LookupExtractorFactoryMapContainer> e : nodeTierLookups.entrySet()) {
+        String name = e.getKey();
+        LookupExtractorFactoryMapContainer lookupToBe = e.getValue();
+
+        LookupExtractorFactoryMapContainer current = lookupsState.getToLoad().get(name);
+        if (current == null) {
+          current = lookupsState.getCurrent().get(name);
+        }
+
+        if (current == null || lookupToBe.replaces(current)) {
+          toLoad.put(name, lookupToBe);
+        }
       }
-      started = false;
-      executorService.shutdownNow();
-      final ListenableScheduledFuture backgroundManagerFuture = this.backgroundManagerFuture;
-      this.backgroundManagerFuture = null;
-      if (backgroundManagerFuture != null && !backgroundManagerFuture.cancel(true)) {
-        LOG.warn("Background lookup manager thread could not be cancelled");
+
+      Set<String> toDrop = new HashSet<>();
```

Comment:
I read this block several times but it still doesn't make sense to me. Please extract it to a method and explain it in a generous comment.